### PR TITLE
Working through liquidation

### PR DESCRIPTION
**Pull request by ith-harvey**
Filled out `kick` & `take` "happy path".
- Implemented accounting for happy path
- Added `Queue` class to track loans in auction

### CODE REVIEW

File: src/base/ScaledPool.sol [hunk 1]
```diff
@@ -18,6 +18,8 @@ import { BucketMath }     from "../libraries/BucketMath.sol";
 import { Maths }          from "../libraries/Maths.sol";
 import { Heap }           from "../libraries/Heap.sol";
 
+import "@std/console.sol";
```
Comment (EdNoepel) [inline src/base/ScaledPool.sol]: remove before flight

File: src/erc20/ERC20Pool.sol [hunk 2]
```diff
@@ -275,71 +278,139 @@ contract ERC20Pool is IERC20Pool, ScaledPool {
...
+        // If recollateralized remove loan from auction
+        if (_borrowerCollateralization(borrower.debt, borrower.collateral, lup) >= Maths.WAD) {
+            _removeAuction(borrower_);
+        }
 
-        // Execute arbitrary code at msg.sender address, allowing atomic conversion of asset
-        msg.sender.call(swapCalldata_);
+        //// TODO: implement flashloan functionality
+        //// Flash loan full amount to liquidate to borrower
+        //collateral().safeTransfer(msg.sender, collateralToPurchase);
```
Comment (EdNoepel) [inline src/erc20/ERC20Pool.sol]: Why is this commented out?  Previous implementation transferred the collateral and then made the external call, as expected.

File: src/erc20/ERC20Pool.sol [hunk 3]
```diff
@@ -275,71 +278,139 @@ contract ERC20Pool is IERC20Pool, ScaledPool {
...
-        _repayDebt(borrower_, quoteTokenReturnAmount);
+        //// Get current swap price
+        //uint256 quoteTokenReturnAmount = _getQuoteTokenReturnAmount(uint256(liquidation.kickTime), uint256(liquidation.referencePrice), collateralToPurchase);
 
-        emit Take(borrower_, amount_, collateralToPurchase, 0);
+        emit Take(borrower_, amount, Maths.wdiv(amount, price), 0);
+        quoteToken().safeTransferFrom(msg.sender, address(this), amount / quoteTokenScale);
     }
 
+    // TODO: remove this method, it is for testing only.
```
Comment (EdNoepel) [inline src/erc20/ERC20Pool.sol]: It'd be nice if `foundry` provided a way for us to test internal functions.  I see this is used by `_assertAuction`, so let's determine a long-term solution before merging.

File: src/erc20/ERC20Pool.sol [hunk 4]
```diff
@@ -414,6 +486,16 @@ contract ERC20Pool is IERC20Pool, ScaledPool {
         );
     }
 
+    function liquidationInfo(address borrower_) external view override returns (uint128, uint256, uint256, uint256) {
+
+        return (
+            liquidations[borrower_].kickTime,            // accrued debt (WAD)
+            liquidations[borrower_].referencePrice,      // deposited collateral including encumbered (WAD)
+            liquidations[borrower_].bondFactor,      // MOMP / inflator, used in neutralPrice calc (WAD)
+            liquidations[borrower_].bondSize // used to calculate pending interest (WAD)
```
Comment (EdNoepel) [inline src/erc20/ERC20Pool.sol]: nit: comment alignment

File: src/erc20/interfaces/IERC20Pool.sol [hunk 5]
```diff
@@ -108,6 +108,16 @@ interface IERC20Pool is IScaledPool {
...
 
+    /**
+     *  @notice Auction is not in the auction Queue.
+     */
+    error AuctionNotActive();
+
+    /**
+     *  @notice Auction is already in the auction Queue.
+     */
+    error AuctionIsActive();
```
Comment (EdNoepel) [inline src/erc20/interfaces/IERC20Pool.sol]: These errors should be in `IScaledPool`; we can reuse for fungible and NFT pools.

File: src/erc20/interfaces/IERC20Pool.sol [hunk 6]
```diff
@@ -135,19 +145,18 @@ interface IERC20Pool is IScaledPool {
      *  @param  borrower_  Address of the borrower.
      *  @return kickTime            Time the liquidation was initiated.
      *  @return referencePrice      Highest Price Bucket at time of liquidation.
-     *  @return remainingCollateral Amount of collateral which has not yet been taken.
-     *  @return remainingDebt       Amount of debt which has not been covered by the liquidation.
+     *  @return bondFactor     BondFactor
+     *  @return bondSize       Bond
```
Comment (EdNoepel) [inline src/erc20/interfaces/IERC20Pool.sol]: nit: doco alignment

File: src/erc20/interfaces/IERC20Pool.sol [hunk 7]
```diff
@@ -168,16 +177,16 @@ interface IERC20Pool is IScaledPool {
...
     /**
      *  @notice Maintains the state of a liquidation.
-     *  @param  kickTime            Time the liquidation was initiated.
-     *  @param  referencePrice      Highest Price Bucket at time of liquidation.
-     *  @param  remainingCollateral Amount of collateral which has not yet been taken.
-     *  @param  remainingDebt       Amount of debt which has not been covered by the liquidation.
+     *  @param  kickTime       Time the liquidation was initiated.
+     *  @param  referencePrice Highest Price Bucket at time of liquidation.
+     *  @param  bondFactor     Bond
+     *  @param  bondSize       Bond
```
Comment (EdNoepel) [inline src/erc20/interfaces/IERC20Pool.sol]: doco should explain `bondFactor` and `bondSize`

File: src/libraries/Maths.sol [hunk 8]
```diff
@@ -146,4 +150,16 @@ library Maths {
...
+    /*** Integer Functions ***/
+    /*************************/
+
+    function maxInt(int256 x, int256 y) internal pure returns (int256) {
+        return x >= y ? x : y;
+    }
+
+    function minInt(int256 x, int256 y) internal pure returns (int256) {
+        return x <= y ? x : y;
+    }
```
Comment (EdNoepel) [inline src/libraries/Maths.sol:164]: We already have `max` and `min` functions which work just fine with unscaled integers.

File: src/_test/ERC20Pool/ERC20PoolGasLoadTest.t.sol [hunk 9]
```diff
@@ -185,7 +182,7 @@ contract ERC20PoolGasLoadTest is ERC20HelperContract {
         for (uint256 i; i < count_; i++) {
             address lender = address(uint160(uint256(keccak256(abi.encodePacked(i, 'lender')))));
 
-            _mintQuoteAndApproveTokens(lender, 200_000 * 1e18);
+            _mintQuoteAndApproveTokens(lender, 300_000_000_000 * 1e18);
```
Comment (EdNoepel) [inline src/_test/ERC20Pool/ERC20PoolGasLoadTest.t.sol]: Why did we suddenly need more tokens for the gas load test?

File: src/_test/ERC20Pool/ERC20ScaledPoolLiquidations.t.sol [hunk 10]
```diff
@@ -4,111 +4,369 @@ pragma solidity 0.8.14;
...
+                bpf:            0,
+                referencePrice: 0,
+                bondFactor:     0,
+                bondSize:       0,
+                next:           address(0),
+                active:         false
+            })
+        );
+
+        //TODO: assert lender state
```
Comment (EdNoepel) [inline src/_test/ERC20Pool/ERC20ScaledPoolLiquidations.t.sol]: Several TODOs in this test...will they be handled with a future PR?

**Review by EdNoepel — CHANGES_REQUESTED** [review]
Some comments inline.  Curious why there were no updates to NFT pool.  The important bit here is the liquidation queue.

File: src/_test/ERC20Pool/ERC20PoolGasLoadTest.t.sol [hunk 11]
```diff
@@ -138,7 +138,6 @@ contract ERC20PoolGasLoadTest is ERC20HelperContract {
 
         assertEq(_loansCount(), LOANS_COUNT);
     }
-
```
Comment (EdNoepel) [inline src/_test/ERC20Pool/ERC20PoolGasLoadTest.t.sol:141]: nit: likely unintentional removal of blank line between functions

File: src/base/PoolInfoUtils.sol [hunk 12]
```diff
@@ -131,6 +131,53 @@ contract PoolInfoUtils {
         lup_      = PoolUtils.indexToPrice(lupIndex_);
     }
 
+    /**
+     *  @notice Returns info related to Auction.
```
Comment (EdNoepel) [inline src/base/PoolInfoUtils.sol:135]: Although it's obvious what to pass, for completeness, natspec should include the two input arguments.

File: src/base/Pool.sol [hunk 13]
```diff
@@ -594,6 +637,152 @@ abstract contract Pool is Clone, Multicall, IPool {
...
+
+        uint256 loansCount = loans.count - 1;
+
+        if (borrowerStruct_.debt != 0) {
+
+            // If loan has debt or collateralized auction has debt
+            if (!auctions.isActive(borrower_) || PoolUtils.collateralization(borrowerStruct_.debt, borrowerStruct_.collateral, lup_) >= Maths.WAD) {
+
+                if (auctions.isActive(borrower_)) auctions.remove(borrower_);
+                if (loansCount != 0
```
Comment (EdNoepel) [inline src/base/Pool.sol:751]: Note: I'm working on changing this to `> 10`.  Happy to leave it `!= 0` for this PR.

File: src/base/Pool.sol [hunk 14]
```diff
@@ -594,6 +637,152 @@ abstract contract Pool is Clone, Multicall, IPool {
...
+        Actors.Borrower memory borrowerStruct_,
+        PoolState memory poolState_,
+        uint256 lup_
+        ) internal {
+
+        uint256 loansCount = loans.count - 1;
+
+        if (borrowerStruct_.debt != 0) {
+
+            // If loan has debt or collateralized auction has debt
```
Comment (EdNoepel) [inline src/base/Pool.sol:747]: I cannot reconcile this comment against the code.  The comment should seemingly read `If loan has debt and either no auction is active or the borrower is well-collateralized`.  Have I misunderstood lines 745-748?

File: src/base/Pool.sol [hunk 15]
```diff
@@ -689,6 +878,47 @@ abstract contract Pool is Clone, Multicall, IPool {
         }
     }
 
+    /**
+     *  @notice Calculates the MOMP, most optomistic matching price.
```
Comment (EdNoepel) [inline src/base/Pool.sol]: sp: `optimistic`

File: src/base/Pool.sol [hunk 16]
```diff
@@ -689,6 +878,47 @@ abstract contract Pool is Clone, Multicall, IPool {
...
+                )
+            );
+    }
+    
+    /**
+     *  @notice Calculates the bondFactor and bondSize, to be used in determining the bond when an auction is kicked. 
+     *  @param numLoans_                  Number of loans in the pool.
+     *  @param poolStateAccruedDebt_      Total debt of the pool.
+     *  @param borrowerAccruedDebt_       Total borrower.
+     *  @param borrowerPledgedCollateral_ Total borrower pledged collateral.
```
Comment (EdNoepel) [inline src/base/Pool.sol]: `lup_` param is missing from natspec

File: src/libraries/PoolUtils.sol [hunk 17]
```diff
@@ -143,6 +144,63 @@ library PoolUtils {
...
+        return PRBMathSD59x18.mul(int256(bondFactor_), _sign(neutralPrice - int256(price_)));
+    }
+
+    /**
+     *  @notice utility function for calculating sign of a value.
+     *  @dev Called in _bpf.
+     *  @param val_   Value to calculate sign on.
+     *  @return sign_ Sign of value.
+     */
+    function _sign(int256 val_) private pure returns (int256) {
```
Comment (EdNoepel) [inline src/libraries/PoolUtils.sol]: Does this belong in `Maths` library?

**Review by EdNoepel — COMMENTED** [review]
Quite happy with this, especially relocating `kick` to `Pool.sol`.  Most of my notes involve documentation.  Note the most recent commit indicates `ERC721Pool is 28787 bytes (over size limit of 24576 bytes).`, which needs to be addressed before merging to develop.

File: src/libraries/PoolUtils.sol [hunk 18]
```diff
@@ -143,6 +144,63 @@ library PoolUtils {
...
+        uint256 collateral_,
+        uint256 mompFactor_,
+        uint256 inflatorSnapshot_,
+        uint256 bondFactor_,
+        uint256 price_
+    ) internal pure returns (int256) {
+        int256 thresholdPrice = int256(Maths.wdiv(debt_, collateral_));
+        int256 neutralPrice = int256(Maths.wmul(mompFactor_, inflatorSnapshot_));
+ 
+        if (thresholdPrice <= neutralPrice) {
```
Comment (mattcushman) [inline src/libraries/PoolUtils.sol]: should be < not <=- to avoid division by zero

**Comment by ith-harvey** [comment]
> Quite happy with this, especially relocating `kick` to `Pool.sol`. Most of my notes involve documentation. Note the most recent commit indicates `ERC721Pool is 28787 bytes (over size limit of 24576 bytes).`, which needs to be addressed before merging to develop.

Working with @grandizzy on a refactor fix that will hopefully solve our problem :crossed_fingers:

**Comment by ith-harvey** [comment]
@grandizzy used this as a guide on his large refactor PRs... this code was included in those PRs, closing.

**Pull request closed without merging**